prefect_azure/blob_storage.py

```diff
@@ -177,3 +189,472 @@ def example_blob_storage_list_flow():
+    _documentation_url = "https://prefecthq.github.io/prefect-azure/blob_storage/#prefect_azure.blob_storabe.AzureBlobStorageContainer"  # noqa
+
+    container_name: str = Field(
+        default=..., description="The name of a Azure Blob Storage container."
+    )
+    credentials: AzureBlobStorageCredentials = Field(
+        default_factory=AzureBlobStorageCredentials,
+        description="The credentials to use for authentication with Azure.",
+    )
+    base_folder: Optional[str] = Field(
+        default=None,
+        description=(
+            "A base path to a folder within the container to use "
+            "for reading and writing objects."
+        ),
+    )
+
+    def _get_path_relative_to_base_folder(self, path: Optional[str] = None) -> str:
+        if path is None and self.base_folder is None:
+            return ""
+        if path is None:
+            return self.base_folder
+        if self.base_folder is None:
+            return path
+        return (Path(self.base_folder) / Path(path)).as_posix()
+
+    @sync_compatible
+    async def download_folder_to_path(
+        self,
+        from_folder: str,
+        to_folder: Union[str, Path],
+        **download_kwargs: Dict[str, Any],
+    ) -> Coroutine[Any, Any, Path]:
+        """Download a folder from the container to a local path.
+
+        Args:
+            from_folder: The folder path in the container to download.
+            to_folder: The local path to download the folder to.
+            **download_kwargs: Additional keyword arguments passed into
+                `BlobClient.download_blob`.
+
+        Returns:
+            The local path where the folder was downloaded.
+
+        Example:
+            Download the contents of container folder `folder` from the container
+                to the local folder `local_folder`:
+
+            ```python
+            from prefect_azure import AzureBlobStorageCredentials
+            from prefect_azure.blob_storage import AzureBlobStorageContainer
+
+            credentials = AzureBlobStorageCredentials(
+                connection_string="connection_string",
+            )
+            block = AzureBlobStorageContainer(
+                container_name="container",
+                credentials=credentials,
+            )
+            block.download_folder_to_path(
+                from_folder="folder",
+                to_folder="local_folder"
+            )
+            ```
+        """
+        self.logger.info(
+            "Downloading folder from container %s to path %s",
+            self.container_name,
+            to_folder,
+        )
+        full_container_path = self._get_path_relative_to_base_folder(from_folder)
+        async with self.credentials.get_container_client(
+            self.container_name
+        ) as container_client:
+            async for blob in container_client.list_blobs(
+                name_starts_with=full_container_path
+            ):
```

Comment:
what do you think of a try/except block here? when my container did not exist, the error is fine but maybe could be better? Like maybe we could have the container name or something in it?

```bash
❯ python qa.py
12:16:56.042 | INFO    | prefect.AzureBlobStorageContainer - Downloading folder from container container to path local_folder
Traceback (most recent call last):
  File "/Users/bean/code-oss/prefect-azure/qa.py", line 14, in <module>
    container.download_folder_to_path(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/utilities/asyncutils.py", line 259, in coroutine_wrapper
    return call()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 431, in __call__
    return self.result()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 317, in result
    return self.future.result(timeout=timeout)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 178, in result
    return self.__get_result()
  File "/Users/bean/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 388, in _run_async
    result = await coro
  File "/Users/bean/code-oss/prefect-azure/prefect_azure/blob_storage.py", line 286, in download_folder_to_path
    async for blob in container_client.list_blobs(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/async_paging.py", line 142, in __anext__
    return await self.__anext__()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/async_paging.py", line 145, in __anext__
    self._page = await self._page_iterator.__anext__()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/async_paging.py", line 94, in __anext__
    self._response = await self._get_next(self.continuation_token)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_list_blobs_helper.py", line 90, in _get_next_cb
    process_storage_error(error)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_shared/response_handlers.py", line 184, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_list_blobs_helper.py", line 83, in _get_next_cb
    return await self._command(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_generated/aio/operations/_container_operations.py", line 1770, in list_blob_flat_segment
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceNotFoundError: The specified container does not exist.
RequestId:3951fc20-301e-0029-6a33-76b96a000000
Time:2024-03-14T17:16:56.3572517Z
ErrorCode:ContainerNotFound
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>ContainerNotFound</Code><Message>The specified container does not exist.
RequestId:3951fc20-301e-0029-6a33-76b96a000000
Time:2024-03-14T17:16:56.3572517Z</Message></Error>
```

Reply:
Added prettier error messages in https://github.com/PrefectHQ/prefect-azure/pull/139/commits/f39d7f05af56fab0aa80d770c6a14d524ed2ec58

Reply:
Nice! For posterity:
```bash
❯ python qa.py
12:56:46.793 | INFO    | prefect.AzureBlobStorageContainer - Downloading folder from container nonexistentcontainer to path local_folder
Traceback (most recent call last):
  File "/Users/bean/code-oss/prefect-azure/prefect_azure/blob_storage.py", line 289, in download_folder_to_path
    async for blob in container_client.list_blobs(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/async_paging.py", line 142, in __anext__
    return await self.__anext__()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/async_paging.py", line 145, in __anext__
    self._page = await self._page_iterator.__anext__()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/async_paging.py", line 94, in __anext__
    self._response = await self._get_next(self.continuation_token)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_list_blobs_helper.py", line 90, in _get_next_cb
    process_storage_error(error)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_shared/response_handlers.py", line 184, in process_storage_error
    exec("raise error from None")   # pylint: disable=exec-used # nosec
  File "<string>", line 1, in <module>
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/aio/_list_blobs_helper.py", line 83, in _get_next_cb
    return await self._command(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/tracing/decorator_async.py", line 77, in wrapper_use_tracer
    return await func(*args, **kwargs)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/storage/blob/_generated/aio/operations/_container_operations.py", line 1770, in list_blob_flat_segment
    map_error(status_code=response.status_code, response=response, error_map=error_map)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/azure/core/exceptions.py", line 164, in map_error
    raise error
azure.core.exceptions.ResourceNotFoundError: The specified container does not exist.
RequestId:e4f03b62-a01e-0076-4838-760d56000000
Time:2024-03-14T17:56:46.9732416Z
ErrorCode:ContainerNotFound
Content: <?xml version="1.0" encoding="utf-8"?><Error><Code>ContainerNotFound</Code><Message>The specified container does not exist.
RequestId:e4f03b62-a01e-0076-4838-760d56000000
Time:2024-03-14T17:56:46.9732416Z</Message></Error>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/bean/code-oss/prefect-azure/qa.py", line 14, in <module>
    container.download_folder_to_path(
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/utilities/asyncutils.py", line 259, in coroutine_wrapper
    return call()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 431, in __call__
    return self.result()
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 317, in result
    return self.future.result(timeout=timeout)
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 178, in result
    return self.__get_result()
  File "/Users/bean/.pyenv/versions/3.10.4/lib/python3.10/concurrent/futures/_base.py", line 391, in __get_result
    raise self._exception
  File "/Users/bean/code-oss/prefect-azure/azurevenv/lib/python3.10/site-packages/prefect/_internal/concurrency/calls.py", line 388, in _run_async
    result = await coro
  File "/Users/bean/code-oss/prefect-azure/prefect_azure/blob_storage.py", line 305, in download_folder_to_path
    raise RuntimeError(
RuntimeError: An error occurred when atempting to download from container nonexistentcontainer: The specified container does not exist.
```